### PR TITLE
Test transport SAR modules

### PR DESCRIPTION
**Pull request by trond-snekvik**
Adds a transport SAR test suite that tests the SAR sender and receiver by mocking endpoints and bearers.

Resolves a range of minor edge case issues, and adding enforcement of invalid behavior that would previously go unchecked.

### CODE REVIEW

**Comment by trond-snekvik** [comment]
cc @soelgary - this adds a few strict checks on transport behavior that could affect the go gateway implementation

File: src/transport/sar/packet.c [hunk 1]
```diff
@@ -42,7 +47,14 @@ int pouch_sar_tx_pkt_decode(const void *buf, size_t buf_len, struct pouch_sar_tx
     }
 
     pkt->seq = bytes[1];
-    pkt->data = &bytes[2];
+    if (buf_len <= POUCH_SAR_TX_PKT_HEADER_LEN)
```
Comment (hasheddan) [inline src/transport/sar/packet.c]: nitpick: we shouldn't ever hit a case where `buf_len < POUCH_SAR_TX_PKT_HEADER_LEN` due to earlier check. Assuming `#define POUCH_SAR_TX_PKT_HEADER_LEN 2`, if `buf_len < POUCH_SAR_TX_PKT_HEADER_LEN`, then the access of the sequence number in the previous line would be invalid (sidenote: potentially worth using a `#define` for the sequence number index rather than just `1`). If somehow we did hit `buf_len < POUCH_SAR_TX_PKT_HEADER_LEN` here, then the following logic would be invalid anyway (though admittedly somewhat safer than than actually accessing the , so I'd lean towards using `==` to more explicitly communicate that this is an "empty" packet.
```suggestion
    if (buf_len == POUCH_SAR_TX_PKT_HEADER_LEN)
```

File: src/transport/sar/receiver.c [hunk 2]
```diff
@@ -103,32 +120,88 @@ int pouch_receiver_recv(struct pouch_receiver *recv, const uint8_t *buf, size_t
...
+        {
+            POUCH_LOG_ERR("Invalid FIN (len %d)", pkt.len);
+            end(recv, false);
+            return -EINVAL;
+        }
+
+        if (recv->state == STATE_IDLE || recv->state == STATE_FAILED)
+        {
+            POUCH_LOG_ERR("Duplicate FIN");
+            return -EINVAL;
```
Comment (hasheddan) [inline src/transport/sar/receiver.c:139]: Just confirming the intention is to _not_ end in this case?

**Review by hasheddan — COMMENTED** [review]
Awesome work @trond-snekvik! A couple of minor comments below.

File: src/transport/sar/packet.c (hunk 1 above)
Comment (trond-snekvik) [reply src/transport/sar/packet.c]: Good observations! Changed.

File: src/transport/sar/receiver.c (hunk 2 above)
Comment (trond-snekvik) [reply src/transport/sar/receiver.c:139]: Yes. If we're already `IDLE` or `FAILED`, we've already done `end()`, and shouldn't do it again (or we're idle because we've never even started). 

There is one exception here though that I have now addressed (and added test coverage for): If the endpoint `start` function fails, we should go back to IDLE. In the reviewed version, we leave it in READY, which is not right.

File: src/transport/sar/packet.c
Comment (trond-snekvik) [inline src/transport/sar/packet.c:39]: Note: There was a bug here that has been fixed in a separate commit. This is the wrong way around - the `POUCH_SAR_TX_PKT_FLAG_IDLE` should be set when we're idling as expected (and not errored out), and the new `bool success = !!(pkt.flags & POUCH_SAR_TX_PKT_FLAG_IDLE);` check in `pouch_receiver_recv` would error out incorrectly, causing the example tests to fail.

**Review by hasheddan — APPROVED** [review]
Nice work @trond-snekvik!

File: src/transport/sar/receiver.c [hunk 3]
```diff
@@ -103,32 +126,88 @@ int pouch_receiver_recv(struct pouch_receiver *recv, const uint8_t *buf, size_t
...
+    }
+
     if (pkt.seq != ((recv->seq + 1) & POUCH_SAR_SEQ_MASK))
     {
-        POUCH_LOG_ERR("OoO RX: %x (last: %x)", pkt.seq, recv->seq);
-        end(recv, false);
-        k_work_reschedule(&recv->work, K_NO_WAIT);  // NACK
-        return -EINVAL;
+        POUCH_LOG_WRN("OoO RX: %x (last: %x)", pkt.seq, recv->seq);
+        // out of order packet - should be ignored
```
Comment (szczys) [inline src/transport/sar/receiver.c:196]: 👏 I had wondered what `OoO` meant before and now I know 😊

**Review by szczys — APPROVED** [review]
Great work @trond-snekvik!!
